FreeRTOS/Test/Target/tests/smp/template/test_name.c

```diff
@@ -0,0 +1,97 @@
+/*
+ * FreeRTOS V202212.00
+ * Copyright (C) 2022 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://www.FreeRTOS.org
+ * https://github.com/FreeRTOS
+ *
+ */
+
+/**
+ * @file test.c
+ * @brief Describe the test here briefly.
+ *
+ * Describe the test here in detail.
+ */
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+
+/* Unity testing includes. */
+#include "unity.h"
+
+/*-----------------------------------------------------------*/
+
+#if ( configNUMBER_OF_CORES < 2 )
+    #error This test is for FreeRTOS SMP and therefore, requires at least 2 cores.
+#endif
+
+#ifndef TEST_CONFIG_H
+    #error test_config.h must be included at the end of FreeRTOSConfig.h.
+#endif
+
+/*-----------------------------------------------------------*/
+
+/* Declare #defines used in this file here. */
+
+/*-----------------------------------------------------------*/
+
+/* Declare local functions used in this file here. */
+
+/*-----------------------------------------------------------*/
+
+/* Declare local variables used in this file here. */
+
+/*-----------------------------------------------------------*/
+
+/* Runs before every test, put init calls here. */
+void setUp( void )
+{
+    /* Create FreeRTOS resources required for the test. */
+}
+/*-----------------------------------------------------------*/
+
+/* Runs after every test, put clean-up calls here. */
+void tearDown( void )
+{
+    /* Delete all the FreeRTOS resources created in setUp. */
+}
```

Comment:
Could `setUp()` and `tearDown()` be common across all tests and defined by each target instead? The main reasons for this is to allow targets to do some memory leak check after each test by calling target specific API. However, this will mean that the test specific setup and teardown will need to be down in the test function itself.

In theory, each target would define something like the following:

```c
static size_t test_start_free_mem;

void setUp( void )
{
    // Get the amount of free memory before test runs
    test_start_free_mem = target_get_cur_free_mem();
}

void tearDown( void )
{
    // Check if test caused a memory leak
    size_t test_end_free_mem = target_get_cur_free_mem();
    TEST_ASSERT_EQUAL(0, test_end_free_mem - test_start_free_mem);
}
```

Reply:
Is it ok to put these memory check on every test runner? I think memory check is not mandatary for all targets.